File: solutions/go/flower-field/1/flower_field.go
package flowerfield

func Annotate(board []string) []string {
	if len(board) == 0 {
		return board
	}

	result := make([]string, len(board))

	for r, row := range board {
		buf := []byte(row)
		for c := range buf {
			if buf[c] == '*' {
				continue
			} else {
				count := 0
				for dr := -1; dr <= 1; dr++ {
					for dc := -1; dc <= 1; dc++ {
						if dr == 0 && dc == 0 {
							continue
						}
						nr, nc := r + dr , c + dc
						if nr >= 0 && nc >= 0 && nr < len(board) && nc < len(board[nr]) {
							if board[nr][nc] == '*' {
								count += 1
							}
						}
					}
				}
				if count > 0 {
					buf[c] = byte('0' + count)
				}
			}
		}
		result[r] = string(buf)
	}

	return result
}

File: solutions/c/flower-field/1/flower_field.c
#include "flower_field.h"
#include <stdlib.h>
#include <string.h>

char **annotate(const char **garden, const size_t rows) {
    if (rows == 0) {
        return NULL;
    }
    char **annotation = malloc((rows + 1) * sizeof(char *));
    for (size_t i = 0; i < rows; i++) {
        annotation[i] = malloc(strlen(garden[i]) + 1);
        strcpy(annotation[i], garden[i]);
    }
    for (size_t i = 0; i < rows; i++) {
        size_t cols = strlen(garden[i]);
        for (size_t j = 0; j < cols; j++) {
            if (garden[i][j] == '*') {
                annotation[i][j] = '*';
            } else {
                int count = 0;
                for (int di = -1; di <= 1; di++) {
                    for (int dj = -1; dj <= 1; dj++) {
                        if (di == 0 && dj == 0) {
                            continue;
                        }
                        int ni = (int)i + di;
                        int nj = (int)j + dj;
                        if (ni >= 0 && ni < (int)rows && nj >= 0 && nj < (int)cols) {
                            if (garden[ni][nj] == '*') {
                                count++;
                            }
                        }
                    }
                }
                if (count > 0) {
                    annotation[i][j] = '0' + count;
                }
            }
        }
    }

    annotation[rows] = NULL;
    return annotation;
}
void free_annotation(char **annotation) {
    for (size_t i = 0; annotation[i] != NULL; i++) {
        free(annotation[i]);
    }
    free(annotation);
}
